### ISSUE TITLE
[bugfix-2.0.x]  Temperature readings unstable

### ISSUE THREAD

**Issue by hamster65** [MarlinFirmware/Marlin#11590]
![2018-08-07 17-37-23](https://user-images.githubusercontent.com/36369379/44307082-7c462e80-a39c-11e8-81e0-9763b3528c36.png)
This is a picture taken about 2 weeks ago. Bed and nozzle temperature gave unstable readings. It stopped as soon as the steppers were activated (for example at the 2 Min mark) and it started once I deactivated them.

![2018-08-19 10-45-55](https://user-images.githubusercontent.com/36369379/44307133-23c36100-a39d-11e8-8617-8e3f6887a389.png)
This is taken today. Both prints failed with a thermal runaway (nozzle). 
This happens after the first layer, when the nozzle temperature is reduced and, at the same time, the layer fan switches on. However, this was never a problem before. 
There seems to be still noise on the extruder temperature reading and at the left, it also seems unstable for the bed sensor. The dips at 17 Min and 7 Min can be ignored (g-code related). I could not find a connection to the stepper state.

Also, sometimes the printer hangs on "E0 heating" forever, and the temperature reading in the lcd is jumping arround. I checked all connections and my ground paths and did not find issues here.
Is anyone observing temperature fluctuations as well?

I'm also wondering if there is some kind of hysteresis for Gcodes like M109 (wait for Hotend temperature). Or does it wait until it stabilizes to the exact temperature? I'd think +- 1°C would be a sufficient margin.

**Comment by kenan-kajkus** [comment]
Did you run PID-Tuning? Looks like not proper PID-values

**Comment by hamster65** [comment]
Yes, but not with the layer fan running because it gets switched of during tuning. 
https://github.com/MarlinFirmware/Marlin/issues/11589

Kp=52,18
Ki=8,97
Kd=75,87

**Comment by comps** [comment]
Maybe related to https://github.com/MarlinFirmware/Marlin/issues/11323 .

**Comment by thinkyhead** [comment]
> I'm also wondering if there is some kind of hysteresis for Gcodes like M109 

```cpp
// Extruder temperature must be close to target for this long before M109 returns success
#define TEMP_RESIDENCY_TIME 10  // (seconds)
#define TEMP_HYSTERESIS 3       // (degC) range of +/- temperatures considered "close" to the target one
#define TEMP_WINDOW     1       // (degC) Window around target to start the residency timer x degC early.

// Bed temperature must be close to target for this long before M190 returns success
#define TEMP_BED_RESIDENCY_TIME 10  // (seconds)
#define TEMP_BED_HYSTERESIS 3       // (degC) range of +/- temperatures considered "close" to the target one
#define TEMP_BED_WINDOW     1       // (degC) Window around target to start the residency timer x degC early.
```

**Comment by kenan-kajkus** [comment]
I think this has nothing to do with histerese, since he says the printer stops because of thermal runaway. Either the component fan blows directly onto the nozzel and cools it down so much that even PID does not help or the PID settings are wrong.

**Comment by AnHardt** [comment]
The parameters for the thermal protections are tunable to tune them. The default values are only suggestions.
You have to tune them for **your** worst case.
Ordering lower temperatures, switching on the part cooling fan, bad fan duct, unisolated heater block, high temperature filtering (delay), bad PID values, all at the same time, is about the worst case. 

Noisy temperature readings **can** help to avoid shutdown because of thermal protection. But in your case you can easily see, the temperature did not reach the target in time.

**Comment by hamster65** [comment]
Thanks thinkyhead! I'll increase the margin. Just today, my printer just waited forever while temperatures were jumping around.
As for the thermal runaway, I think it is a different problem which has nothing to do with the unstable readings.

**Comment by hamster65** [comment]
I swapped my E3D Hotend to a volcano version and ran PID autotune (M303 S210 U C10).
 This is what happens while printing. 
![2018-08-26 18-08-17](https://user-images.githubusercontent.com/36369379/44630284-27398800-a95b-11e8-8efb-1b6fa59561d4.png)
PID values: 
Kp 35.90
Ki 5.22
Kd 61.75

**Comment by hamster65** [comment]
I swapped my E3D Hotend to a volcano version and ran PID autotune (M303 S210 U C10).
 This is what happens while printing. 
![2018-08-26 18-08-17](https://user-images.githubusercontent.com/36369379/44630284-27398800-a95b-11e8-8efb-1b6fa59561d4.png)
PID values: 
Kp 35.90
Ki 5.22
Kd 61.75

**Comment by AnHardt** [comment]
It was noise yesterday and still is today. That can't be a real change in temperature - too fast. No nozzle will heat faster than when full on and not cool faster than when completely off. Check capacitors, shielding, seperating from the stepper/heater cables, ... . Be sure to use one of the branches from after Jul 26 2018 having #11370  or #11336 in.

**Comment by hamster65** [comment]
I changed one cable to 2,5mm². Everything except for the heated bed runs through this fairly long cable. With the new cable, the leds stopped flickering when the nozzle is heating up. 

However, it did not stop the temperature issue and I cannot explain the readings:
![2018-08-28 13-36-47](https://user-images.githubusercontent.com/36369379/44720553-77cdf400-aac7-11e8-98fa-34f6670b65f6.png)
From 13:24 to 13:31, only the heated bed was active. Its running in bang bang mode, so most of the time, it did not draw current. Fans, steppers were disabled. So even when there is almost no current flow, fluctuations occur for both sensors.
When the nozzle is heated to 50°C, the nozzle sensor does not show the fluctuations any more, which again is strange.

Power supply for the arduino mega looks clean, no fluctuations over 14 seconds and also no spikes visible in zoom or when I look at higher frequencies.
![ds2_quickprint1](https://user-images.githubusercontent.com/36369379/44722510-8d461c80-aacd-11e8-9d43-4e846ee459cb.png)

I flashed a minimalistic configuration of the current 2.0 branch. No difference.
